### DISCUSSION TITLE
src-static structure

### DISCUSSION THREAD

**Discussion by zackarno** [OCHA-DAP/hdx-signals#121] [Ideas]
This is not really an issue, but a potential idea. It's likely you already have thought about it and decided against it, so I'm making it a discussion rather than issue.

I'm finding it a little bit tricky to develop the `src-static`  functions and I think the main reason is the way that scripts are mix function definitions with calls to download as well as upload data. I understand that these functions are "supposed" to only be called once, but I think when it comes to testing and developing these they need to be called in various different ways and the current set up is not the easiest (at least for some one with my skill). 

My idea to make this easier would just be to have the `src-static` files just be functions and there implementations occur in a separate folder like `data-raw` (the R pkg convention). This would make allow me to actually source the `src-static` files and functions from anywhere which i cant actually do now for 2 reasons : 1.) box does not like the `-` in src-static (easy fix) , 2. if I fix 1 by changing `-` -> `_` , and run `box::use(./src_static/update_iso_sf)` it srcs the whole file and starts downloading everything.

As writing this I think it's  mainly the implementation of the defined functions (the uploading) that causes issues for me. The downloading to create objects outside of functions seems fine, for example `sf_world` object (`sf_world <- cs$read_az_file("input/un_geodata.geojson")`) doesn't make it harder to develop.

In addition to easing developing/testing (manual/automatic), I think the separation of functions and there implementation could just conceptually makes the repo easier to follow

**Comment by zackarno** [comment]
Here is a just one quick example of how I find it limiting for development. Don't take my example too literally.  Say I want to use `update_iso_sf` functions to create a single global admin 0 geoparquet file (something we had discussed briefly).

Below I've  abstracted part of the `update_` function so that it will return an sf class object.

https://github.com/OCHA-DAP/hdx-signals/blob/afa511e6abb3b362005cc81be12dafd9b794991a/src-static/update_iso3_sf.R#L49-L88

I'd like to simply open a blank script and write (semi-psuedocode)
```
box::use(../src_static/update_iso3_sf[simplify_adm0])
box::use(purrr)
box::use(arrow)

boundaries <- purrr$map(
  .x = all_iso3_codes(),
  .f = simplify_adm0
) |> 
  purrr$list_rbind()

arrow$write_parquet(...)
```

likely i'd want run this as a background process through the terminal since it takes a long time. I don't think its currently possible.

This is just one example, and it probably seems quite specific, but i think its actually quite general to the development process when you might want to refactor and expand functionality

**Comment by caldwellst** [comment]
Yeah, I'm definitely reticent to turn the scripts into functions! However, open to it if absolutely necessary, so let's discuss.

If it were me wanting to, in this case, develop a global map and use one of those functions, to start my testing I would functionalise it and put it in the utilities. Then I could import it in the original script and the new one I'm testing. And if you want to test your new script, you then could use `Rscript new_script.R`.

Not sure how functionalizing each `src-static` script would help the above?

**Reply by caldwellst** [reply]
I'm sure it was just an example though, but side note, please don't create a global admin file ha, as we only create country maps so it would just add extra work to align geometries (maybe) and add a line of code to filter a global dataset to a country file for every map.

**Reply by zackarno** [reply]
ha yeah im not actually making a global file, but as part of my testing to see if the data_sources etc are being added on correctly, I  am looping through all admin codes and creating a temporary admin 0 object with all geoms so i can make sure things are going smoothly... and yeah there are some errors so its useful to do this same type of purrr$map operation shown in the example.


> to start my testing I would functionalise it and put it in the utilities. 

yes this is basically what I am doing, but it means that that every time I want to test, expand, refactor functionality of these functions I have to copy them into a new "utils-type" script.... rather than simply

```
box::use(./src-static/static_script[func_i_want_to_work_on])

#> develop write some test code
```

The above would be possible if the static_scripts were just functions so they could be modularly imported... i think the downloading/loading of certain objects to the script/box environment like `sf_world` is fine and maybe even a smart way to reduce the number of arguments required by functions. It's really just the big final purrr implementations of the functions that make is so they can't /shouldn't really be sourced when developing

**Reply by caldwellst** [reply]
I see what you mean! That makes sense actually. We could just functionalise the final run. I thought you meant the entire script. Maybe in all of the static scripts we could have a function `update()` that does the run.

Or, even easier, we could just have the final run at the end wrapped in `if (!interactive())` so that the actual updates run when not interactively developing.
